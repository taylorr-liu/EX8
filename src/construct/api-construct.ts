import { RestApi, LambdaIntegration, Resource, RestApiProps } from '@aws-cdk/aws-apigateway';
import { Function } from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

export interface ItemsAPIProps extends RestApiProps {
  /**
     * @CURDLambdaFunctions Functions for CURD unspecified items
     */
  generalLambdaFunctions?: CURDLambdaFunctions;
  /**
     * @CURDLambdaFunctions Functions for CURD specified items
     */
  specificLambdaFunctions?: CURDLambdaFunctions;
}

export interface CURDLambdaFunctions {
  /**
     * @Function Function for the GET method
     */
  GET?: Function;

  /**
     * @Function Function for the POST method
     */
  POST?: Function;

  /**
     * @Function Function for the PUT method
     */
  PUT?: Function;

  /**
     * @Function Function for the DELETE method
     */
  DELETE?: Function;
}

export class ItemsAPI extends RestApi {
  /**
     * @Resource The API Resource of general methods.
     */
  public readonly generalMethodEntry: Resource;

  /**
     * @Resource The API Resource of specific methods.
     */
  public readonly specificMethodEntry: Resource;

  /**
     * The API with two kinds of resource, general methods and specific methods.
     * @param scope
     * @param id
     * @param props
     */
  constructor(scope: Construct, id: string, props: ItemsAPIProps) {
    super(scope, id, props);
    this.generalMethodEntry = this.root.addResource('items');
    this.specificMethodEntry = this.generalMethodEntry.addResource('{id}');

    if (props.generalLambdaFunctions) {
      this.setGeneralLambdaFunctions(props.generalLambdaFunctions);
    }
    if (props.specificLambdaFunctions) {
      this.setSpecificLambdaFunctions(props.specificLambdaFunctions);
    }
  }

  /**
     * Set the methods for the API resource.
     * @param resource
     * @param lambdaFuncs
     */
  private setLambdaFunctions(resource: Resource, lambdaFuncs: CURDLambdaFunctions) {
    if (lambdaFuncs.GET) {
      resource.addMethod('GET', new LambdaIntegration(lambdaFuncs.GET));
    }
    if (lambdaFuncs.POST) {
      resource.addMethod('POST', new LambdaIntegration(lambdaFuncs.POST));
    }
    if (lambdaFuncs.PUT) {
      resource.addMethod('PUT', new LambdaIntegration(lambdaFuncs.PUT));
    }
    if (lambdaFuncs.DELETE) {
      resource.addMethod('DELETE', new LambdaIntegration(lambdaFuncs.DELETE));
    }
  }

  /**
     * Set the lambda functions for the general methods.
     * @param lambdaFuncs The lambda functions for the general methods.
     */
  public setGeneralLambdaFunctions(lambdaFuncs: CURDLambdaFunctions) {
    this.setLambdaFunctions(
      this.generalMethodEntry,
      lambdaFuncs,
    );
  }

  /**
     * Set the lambda functions for the specific methods.
     * @param lambdaFuncs The lambda functions for the general methods.
     */
  public setSpecificLambdaFunctions(lambdaFuncs: CURDLambdaFunctions) {
    this.setLambdaFunctions(
      this.specificMethodEntry,
      lambdaFuncs,
    );
  }
}