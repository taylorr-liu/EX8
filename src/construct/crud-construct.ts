import { join } from 'path';
import { Table } from '@aws-cdk/aws-dynamodb';
import { Role } from '@aws-cdk/aws-iam';
import { Function } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { Construct } from '@aws-cdk/core';

export interface CURDConstructProps {
  /**
     * @Table table The DynamoDB table to CURD.
     */
  table: Table;
  /**
     * @Role tableRole The IAM role with the policy to access the previous DynamoBD
     */
  tableRole: Role;
}

/**
 * A construct having CURD lambda functions.
 */
export class CURDConstruct extends Construct {
  /**
     * @Function The lambda function to get a specific item.
     */
  public readonly getOne: Function;
  /**
     * @Function The lambda function to get all items.
     */
  public readonly getAll: Function;
  /**
     * @Function The lambda function to create one item.
     */
  public readonly createOne: Function;
  /**
     * @Function The lambda function to update a specific item.
     */
  public readonly updateOne: Function;
  /**
     * @Function The lambda function to delete a specific item.
     */
  public readonly deleteOne: Function;

  /**
     * A construct having CURD lambda functions.
     * @param scope
     * @param id
     * @param props The props containing the DynamoDB Table to access and the IAM Role with accessibility.
     */
  constructor(scope: Construct, id: string, props: CURDConstructProps) {
    super(scope, id);

    this.getOne = new NodejsFunction(this, 'getOneItemFunction', {
      entry: `${process.env.APPS_PATH}/data/getone/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.getAll = new NodejsFunction(this, 'getAllItemsFunction', {
      entry: `${process.env.APPS_PATH}/data/getall/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.createOne = new NodejsFunction(this, 'createItemFunction', {
      entry: `${process.env.APPS_PATH}/data/create/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.updateOne = new NodejsFunction(this, 'updateItemFunction', {
      entry: `${process.env.APPS_PATH}/data/update/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.deleteOne = new NodejsFunction(this, 'deleteItemFunction', {
      entry: `${process.env.APPS_PATH}/data/delete/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });
  }
}