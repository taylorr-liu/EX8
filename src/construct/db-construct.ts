import { Table, AttributeType } from '@aws-cdk/aws-dynamodb';
import { PolicyStatement, Effect } from '@aws-cdk/aws-iam';
import { Construct, RemovalPolicy } from '@aws-cdk/core';

export interface ItemTableProps {
  tableName: string;
  primaryKeyName: string;
}

export class ItemTable extends Table {
  constructor(scope: Construct, id: string, props: ItemTableProps) {
    super(scope, id, {
      partitionKey: {
        name: props.primaryKeyName,
        type: AttributeType.STRING,
      },
      tableName: props.tableName,
      removalPolicy: RemovalPolicy.DESTROY,
    });
  }
  getAccessPolicy(): PolicyStatement {
    return new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        'dynamodb:PutItem',
        'dynamodb:UpdateItem',
        'dynamodb:DeleteItem',
        'dynamodb:GetItem',
        'dynamodb:Query',
        'dynamodb:Scan',
      ],
      resources: [this.tableArn],
    });
  }
}