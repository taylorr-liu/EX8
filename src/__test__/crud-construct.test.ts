import '@aws-cdk/assert/jest';
import * as path from 'path';
import { Table, AttributeType } from '@aws-cdk/aws-dynamodb';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { App, Stack } from '@aws-cdk/core';
import { CURDConstruct } from '../construct/crud-construct';

test('curd-construct', () => {
  const app = new App();
  const stack = new Stack(app, 'test');

  process.env.APPS_PATH = path.resolve(__dirname, '..', 'apps', 'src');
  new CURDConstruct(stack, 'CURDConsturct', {
    table: new Table(stack, 'test-table', {
      partitionKey: {
        name: 'items',
        type: AttributeType.STRING,
      },
    }),
    tableRole: new Role(stack, 'IAMFunctionRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
    }),
  });
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
  expect(stack).toCountResources('AWS::Lambda::Function', 5);

});