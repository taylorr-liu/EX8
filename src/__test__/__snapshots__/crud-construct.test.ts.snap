// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`curd-construct 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565ArtifactHashF8724CB9": Object {
      "Description": "Artifact hash for asset \\"5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565\\"",
      "Type": "String",
    },
    "AssetParameters5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565S3Bucket00F98B11": Object {
      "Description": "S3 bucket for asset \\"5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565\\"",
      "Type": "String",
    },
    "AssetParameters5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565S3VersionKey3DE3FA9D": Object {
      "Description": "S3 key for asset version \\"5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565\\"",
      "Type": "String",
    },
    "AssetParameters958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7ArtifactHash9F415277": Object {
      "Description": "Artifact hash for asset \\"958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7\\"",
      "Type": "String",
    },
    "AssetParameters958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7S3BucketF90A5F2A": Object {
      "Description": "S3 bucket for asset \\"958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7\\"",
      "Type": "String",
    },
    "AssetParameters958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7S3VersionKey789B059F": Object {
      "Description": "S3 key for asset version \\"958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7\\"",
      "Type": "String",
    },
    "AssetParameterse4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90daArtifactHash327F0BEE": Object {
      "Description": "Artifact hash for asset \\"e4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90da\\"",
      "Type": "String",
    },
    "AssetParameterse4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90daS3Bucket611694C2": Object {
      "Description": "S3 bucket for asset \\"e4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90da\\"",
      "Type": "String",
    },
    "AssetParameterse4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90daS3VersionKey54C309E7": Object {
      "Description": "S3 key for asset version \\"e4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90da\\"",
      "Type": "String",
    },
    "AssetParametersefb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23ArtifactHash71C6A1FF": Object {
      "Description": "Artifact hash for asset \\"efb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23\\"",
      "Type": "String",
    },
    "AssetParametersefb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23S3Bucket049FFAA5": Object {
      "Description": "S3 bucket for asset \\"efb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23\\"",
      "Type": "String",
    },
    "AssetParametersefb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23S3VersionKey92E97921": Object {
      "Description": "S3 key for asset version \\"efb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23\\"",
      "Type": "String",
    },
    "AssetParametersf58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeabArtifactHashBDCB63B8": Object {
      "Description": "Artifact hash for asset \\"f58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeab\\"",
      "Type": "String",
    },
    "AssetParametersf58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeabS3Bucket0558EA48": Object {
      "Description": "S3 bucket for asset \\"f58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeab\\"",
      "Type": "String",
    },
    "AssetParametersf58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeabS3VersionKey0CCB65A8": Object {
      "Description": "S3 key for asset version \\"f58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeab\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CURDConsturctcreateItemFunction216B2712": Object {
      "DependsOn": Array [
        "IAMFunctionRole12477571",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565S3Bucket00F98B11",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565S3VersionKey3DE3FA9D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5e6cc11d6fc5f756318c4da51204aaf93387247aa3c7af747583284c41a84565S3VersionKey3DE3FA9D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "PRIMARY_KEY": "itemId",
            "TABLE_NAME": Object {
              "Ref": "testtableB6F02C77",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "IAMFunctionRole12477571",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CURDConsturctdeleteItemFunctionCD4C0122": Object {
      "DependsOn": Array [
        "IAMFunctionRole12477571",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersf58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeabS3Bucket0558EA48",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeabS3VersionKey0CCB65A8",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf58bc90eb68df3f67189b858e14016cf78de5650c1b169b272646808ff05eeabS3VersionKey0CCB65A8",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "PRIMARY_KEY": "itemId",
            "TABLE_NAME": Object {
              "Ref": "testtableB6F02C77",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "IAMFunctionRole12477571",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CURDConsturctgetAllItemsFunction7EDB612C": Object {
      "DependsOn": Array [
        "IAMFunctionRole12477571",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7S3BucketF90A5F2A",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7S3VersionKey789B059F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters958462d32024951aff726d39d4968c6323b0897f175df8d4f2d097b05209cac7S3VersionKey789B059F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "PRIMARY_KEY": "itemId",
            "TABLE_NAME": Object {
              "Ref": "testtableB6F02C77",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "IAMFunctionRole12477571",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CURDConsturctgetOneItemFunction07B0C44A": Object {
      "DependsOn": Array [
        "IAMFunctionRole12477571",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterse4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90daS3Bucket611694C2",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90daS3VersionKey54C309E7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse4071692f63cf9a43d30bb4623ac709dd72aca760c602359b89ac2fd3ece90daS3VersionKey54C309E7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "PRIMARY_KEY": "itemId",
            "TABLE_NAME": Object {
              "Ref": "testtableB6F02C77",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "IAMFunctionRole12477571",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CURDConsturctupdateItemFunctionE26FAD21": Object {
      "DependsOn": Array [
        "IAMFunctionRole12477571",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersefb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23S3Bucket049FFAA5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersefb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23S3VersionKey92E97921",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersefb75661c75bed49d13dab001dc7f7676c4bdb92e40b443538f6d1a63d74ca23S3VersionKey92E97921",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "PRIMARY_KEY": "itemId",
            "TABLE_NAME": Object {
              "Ref": "testtableB6F02C77",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "IAMFunctionRole12477571",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "IAMFunctionRole12477571": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testtableB6F02C77": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "items",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "items",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
